Feature/sync measurement units

FeliGoblin opened the pull request:
Apologies for the issues from the last PR. 🙏

## Proposed change:

Each sensor entity relating to pet feed now has a Volume entity and Weight entity. Separating them like this allows for long-term statistics. 
![Screenshot_1](https://github.com/user-attachments/assets/e1e88033-09dd-4ebf-b77f-848937bf40da)

When a user changes a unit of measurement in Account Settings, the relevant entities will update to reflect their choice. Note that 'cups' are not supported in Home Assistant .
The user can also choose "Update measurement unit of all Petlibro sensor entities" to update the measurement unit of _all_ of their feed, weight and water entities regardless of what units are or aren't changed. This can be useful if the user changes a measurement unit outside of Home Assistant (like in the app).
![Screenshot_20](https://github.com/user-attachments/assets/ff3c6bf0-641f-4b85-a55c-418aaa65bbd2)

These entities also have the values for the 4 Petlibro feed unit options plus a "portion" unit (API "grain") in the extra attributes.
![Screenshot_11](https://github.com/user-attachments/assets/ad873ad6-16e1-4ee7-8443-8bfa70018361)

The Manual Feed Quantity number entities will use the feed unit chosen by the user in their settings. Please see the first comment about how this looks for 'cups'.

Closes #141 #133 #131 #8  (issue)

## Type of change:

- [ ] New device.
- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature or enhancement (non-breaking change which adds functionality).
- [ ] Documentation only.
- [ ] Other (please explain).

## Checklist:

- [ ] If applicable, I have added corresponding documentation changes.
- [x] If applicable, I have reviewed the [feature / enhancement](https://github.com/jjjonesjr33/petlibro/wiki/Feature-&-Enhancement-Guidelines) guidlines before submitting my request.
- [x] If applicable, I have tested my code for new features & regressions on the latest version of Home Assistant.

## Additional notes:

FeliGoblin commented:
For the Manual Feed Quantity number entity when 'cups' are being used as the feed unit..
This one's a bit awkward because a single portion size divides a cup by 12, so 0.08333333333 of a cup, which doesn't work very nicely in a number entity.

We've got a few options..

**2DP:**
Instead of dividing by 12, we could divide by 20 and have increments of 0.05. The actual stored value will round to the nearest 1/12, but the slider just looks a bit cleaner.

**3DP:**
12 steps per cup, and starts at 0.083. The number on the slider won't be able to show nice halves (like 0.5) and wholes (like 1) like 2DP does. Instead it'll be 0.498 for half a cup, 0.996 for 1 cup, 3.984 for 4 cups etc. Actual stored value is also corrected though. 

**Just use the big exact decimal:**
Just have the slider show 0.08333333333, 2.16666666666667, 2.25, 2.5, 3.66666666666667 etc.
This is how the PR currently is.

**Portions:**
Use portion sizes, this is how it worked before with a "/12 cups" unit of measurement. Looks a bit weird when going above 1 cup, "34/12 cups" etc, but it works. Could also change the unit of measurement to just be "portions".

**Fractional values as the unit of measurement:**
The slider numbers can be anything, but the UoM shows the actual value as a fraction. (³/₁₂ cups, 2 ²/₁₂ cups etc). Can look nice in cards if the user hides the entity state and only shows the UoM, but otherwise it looks bizarre. 

**Dropdown list**
Or just have a dropdown list.

Examples:

https://github.com/user-attachments/assets/fe1e64a3-aeb1-4906-9960-8dfa298ddf72



Another option I don't have in the video is to have the slider be just whole cups from 0 to 4, and then have a second number entity (which only appears for people using cups) for the fractions 0 to 11

Personally I'd probably go for exact decimals, or one of the portions, or the extra fractions entity, but what do you think? Any other ideas I haven't thought of?

@jjjonesjr33 @C4-Dimitri @tube0013

C4-Dimitri commented:
@FeliGoblin, not a cup user so won't comment on that.

But everything else looks good to me. 1 question, have you thought about and implemented it for the number entitiy for the fountains? Low level threshold i think its called. Thats currently defaulted to mL as well.

FeliGoblin commented:
> @FeliGoblin, not a cup user so won't comment on that.
> 
> But everything else looks good to me. 1 question, have you thought about and implemented it for the number entitiy for the fountains? Low level threshold i think its called. Thats currently defaulted to mL as well.

I did, and I just noticed I made a mistake in it 🤦 Fixing it now lol

tube0013 commented:
I'm inclined to like the portions, but I have no problem knowing 18 portions is a cup and a half for example. If this is explained in the docs briefly I think it would cover it. 

I was surprised when I initially got the device that it did it this way. I'm not opposed to switching to grams either. Silly that cups is volume and grams is weight anyway. Rumor has it in the US food serving sizes will be changing to grams sometime. Volume is useless without density. I know when I had ChatGPT update it so I could dispense more it just kept the /12.

jjjonesjr33 commented:
@FeliGoblin The portions example is definitely more user-friendly and simple to grasp, but I’m leaning toward the dropdown. It gives exact values, removes the guesswork, and makes it clear how much each portion actually represents.

The real question is, if a user switches from cups how would that show in the dropdown? Or does it change?

FeliGoblin commented:
> @FeliGoblin The portions example is definitely more user-friendly and simple to grasp, but I’m leaning toward the dropdown. It gives exact values, removes the guesswork, and makes it clear how much each portion actually represents.
> 
> The real question is, if a user switches from cups how would that show in the dropdown? Or does it change?

I'm thinking the select entity would be disabled and hidden, and a number entity enabled and un-hidden to replace it. Or vice versa if switching _to_ cups.
I haven't coded it up yet but it should be fairly simple I think

C4-Dimitri commented:
> select entity would be disabled and hidden, and a number entity enabled and un-hidden to replace it.

Didn't know we could do that via the integration, if thats the case this sounds like an ideal solution.

FeliGoblin commented:
> 
> I'm thinking the select entity would be disabled and hidden, and a number entity enabled and un-hidden to replace it. Or vice versa if switching _to_ cups. I haven't coded it up yet but it should be fairly simple I think

So to do this I need to reload the integration after re-enabling an entity, which is fine but it does mean the user has to wait a minute for the data to populate again.

Speaking of which... and just to make sure I'm not breaking anything, is there a reason we have the integration waiting a minute for the first refresh before getting all the entity data? We could have the data refreshed immediately when the integration loads if we remove the `self.last_refresh_times[device_sn] = datetime.utcnow()` from `load_devices(self)`. As it is now, the `await hub.coordinator.async_config_entry_first_refresh()` in __init__.py isn't really doing anything because the `last_refresh_times` was just set even though the data hasn't been refreshed yet, it's only called `list_devices()` at this point.
@jjjonesjr33

jjjonesjr33 commented:
@FeliGoblin It needs to wait before making the first pull at boot so the device has time to come online and be confirmed as available. If I remember right, there’s a delay between logging into the account, fetching the data, and populating the entries since the API is split across multiple calls — `realInfo`, `baseInfo`, and `grainStatus`

C4-Dimitri commented:
@FeliGoblin, 1 more question, have you made each "amount" in the below go through the same logic? If not, we should, so "amount volume" & "amount weight" and then the correct UOM:
![capture_temp](https://github.com/user-attachments/assets/2fdf3845-24c1-4546-9fa2-1b44f16a5714)

FeliGoblin commented:
Alright, updated the PR with the following:

The Manual Feed Quantity number entities will use the feed unit chosen by the user in their settings. 
![Screenshot_7](https://github.com/user-attachments/assets/2b802f34-a064-43e8-902c-7ebd07f4d58d)

If the user has their feed unit set to cups, the Manual Feed Quantity entity will be a Select (dropdown) instead.
![Screenshot_5](https://github.com/user-attachments/assets/ec84a98f-3d48-4c72-a28f-a8a43fa39133)

Also added an option to switch the Manual Feed Quantity entity to use 'portions' instead of the measurement unit.
![Screenshot_1](https://github.com/user-attachments/assets/ef887e7e-4754-465b-ac7b-a70c5995e657)
![Screenshot_2](https://github.com/user-attachments/assets/bd1fff53-0caa-4005-aa9d-c203ac459210)
![Screenshot_4](https://github.com/user-attachments/assets/cbddd6c6-58f5-4ae5-bfe5-3dfee78dea5c)

Also replaced "Current Weight" with "Remaining Water" for fountains. Slightly renamed the percentage entity from "Remaining Water" to "Remaining Water (%)" and changed the icon from "mdi:water" to "mdi:water-percent"
![Screenshot_8](https://github.com/user-attachments/assets/42547a05-9aae-4dd8-a56e-39848aaeafc4)

> @FeliGoblin, 1 more question, have you made each "amount" in the below go through the same logic? If not, we should, so "amount volume" & "amount weight" and then the correct UOM: 

And yep, that's in there :)

FeliGoblin commented:
It's ready for merging if you're happy with it @jjjonesjr33  :)

jjjonesjr33 reviewed (approved):
Approved, let's give it a test.

jjjonesjr33 commented:
Linking https://github.com/jjjonesjr33/petlibro/issues/173
